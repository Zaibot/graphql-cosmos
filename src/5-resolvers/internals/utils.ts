import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse(
  typename: string,
  database: string,
  container: string,
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<{ id: string }>>>,
  count: Lazy<Promise<FeedResponse<number>>>
): GraphQLCosmosPageOutput {
  return {
    get cursor() {
      return Promise.resolve(cursor)
    },
    get nextCursor() {
      return query().then((x) => x.continuationToken)
    },
    get page() {
      return query().then((x) => x.resources.map(wrapSingleSourceDescriptor(typename, database, container)))
    },
    get total() {
      return count().then((x) => x.resources[0])
    },
  }
}

export function wrapSingleSourceDescriptor(typename: string, database: string, container: string) {
  // TODO, use refSingle of datasource
  return (item: { id: string }) =>
    SourceDescriptor.withDescriptor(item, {
      kind: `Single`,
      database,
      container,
      typename,
      id: item.id,
    })
}
