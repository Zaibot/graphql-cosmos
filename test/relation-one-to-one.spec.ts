import gql from 'graphql-tag'
import { printSchemaWithDirectives } from 'graphql-tools'
import { createUnitTestContext } from './utils'

describe(`One to one`, () => {
  const dummyTypeDefs = gql`
    type Query {
      lefts: [Left] @cosmos(database: "Test", container: "Lefts")
    }

    type Left {
      id: ID!
      right: Right @cosmos(database: "Test", container: "Rights", ours: "rightId")
    }

    type Right {
      id: ID!
    }
  `

  const responses = {
    Lefts: {
      'SELECT VALUE COUNT(1) FROM c': [1],
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `l` }],
      'SELECT c.id, c.rightId FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=l)': [
        { id: `l`, rightId: `r` },
      ],
    },
  }

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })

  it(`should be retrieve all items`, async () => {
    const result = await uc.execute(`query { lefts { total page { __typename id right { __typename id } } } }`)

    expect(result).toEqual({
      data: {
        lefts: {
          total: 1,
          page: [
            {
              __typename: 'Left',
              id: `l`,
              right: { __typename: 'Right', id: `r` },
            },
          ],
        },
      },
    })
  })
})
