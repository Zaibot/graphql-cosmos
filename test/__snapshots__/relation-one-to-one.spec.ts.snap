// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`One to one expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Lefts",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Rights",
          "cosmos": true,
          "database": "Test",
          "fieldname": "right",
          "ours": "rightId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Right",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Left",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Lefts",
          "cosmos": true,
          "database": "Test",
          "fieldname": "lefts",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Left",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
    Object {
      "container": "Rights",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Right",
    },
  ],
}
`;

exports[`One to one expects schema to remain the same 1`] = `
"schema {
  query: Query
}

type Query {
  lefts(cursor: String, limit: Int): LeftPage!
}

type Left {
  id: ID!
  right: Right
}

type Right {
  id: ID!
}

type LeftPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Left]!
}
"
`;
